### PR TITLE
Update CSS Color parsing and serialization to match the current spec

### PR DESCRIPTION
**Pull request by weinig**
#### 8913e90d9b69466796a2c4c7114f425e95bfa964
<pre>
Update CSS Color parsing and serialization to match the current spec
<a href="https://bugs.webkit.org/show_bug.cgi?id=245276">https://bugs.webkit.org/show_bug.cgi?id=245276</a>

Reviewed by Darin Adler.

* LayoutTests/imported/w3c/web-platform-tests/css/css-color/animation/color-interpolation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-contrast-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-contrast-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-mix-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-mix-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-mixed-num-pct-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/relative-color-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/relative-color-invalid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/relative-color-valid-expected.txt:
Update test results. Some of these now report as failing, but the tests are incorrect and need
updating in WPT.

* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::normalizeLightnessPercent):
(WebCore::CSSPropertyParserHelpers::normalizeABPercent):
(WebCore::CSSPropertyParserHelpers::normalizeChromaPercent):
(WebCore::CSSPropertyParserHelpers::normalizeXYZPercent):
(WebCore::CSSPropertyParserHelpers::normalizeRGBPercent):
Helpers to implement the percent scaling now required by the spec.

(WebCore::CSSPropertyParserHelpers::parseLabParameters):
(WebCore::CSSPropertyParserHelpers::parseRelativeLabParameters):
(WebCore::CSSPropertyParserHelpers::parseNonRelativeLabParameters):
(WebCore::CSSPropertyParserHelpers::parseLCHParameters):
(WebCore::CSSPropertyParserHelpers::parseRelativeLCHParameters):
(WebCore::CSSPropertyParserHelpers::parseNonRelativeLCHParameters):
(WebCore::CSSPropertyParserHelpers::parseColorFunctionForRGBTypes):
(WebCore::CSSPropertyParserHelpers::parseColorFunctionForXYZTypes):
(WebCore::CSSPropertyParserHelpers::parseRelativeColorFunctionForXYZTypes):
Update parsing to support both numeric and percent for non-hue components.

* Source/WebCore/platform/graphics/ColorConversion.cpp:
(WebCore::OKLab&lt;float&gt;&gt;::convert):
(WebCore::WhitePoint::D65&gt;&gt;::convert):
Remove now unneeded percent transforms since we now store
lightness in its numeric form.

* Source/WebCore/platform/graphics/ColorSerialization.cpp:
(WebCore::serializationOfLabFamilyForCSS):
(WebCore::serializationForCSS):
(WebCore::percentageComponent): Deleted.
(WTF::StringTypeAdapter&lt;WebCore::PercentageComponent&gt;::StringTypeAdapter): Deleted.
(WTF::StringTypeAdapter&lt;WebCore::PercentageComponent&gt;::length const): Deleted.
(WTF::StringTypeAdapter&lt;WebCore::PercentageComponent&gt;::is8Bit const): Deleted.
(WTF::StringTypeAdapter&lt;WebCore::PercentageComponent&gt;::writeTo const): Deleted.
(WTF::StringTypeAdapter&lt;WebCore::PercentageComponent&gt;::buffer const): Deleted.
Update serialization to use the numeric form for lightness in lab/lch/oklab/oklch.
The whole family now has the same serialization scheme so can be hoisted into
a new helper, serializationOfLabFamilyForCSS.

Canonical link: <a href="https://commits.webkit.org/254688@main">https://commits.webkit.org/254688@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/24ab603fdce3855356f99a7a35f84b6185ad4384

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/89897 "16 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/34443 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/20558 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/99232 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/156280 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/93904 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/32935 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/28353 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/82229 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/93552 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/95543 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/26155 "Passed tests") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/76698 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/82229 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/81031 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/20558 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/82229 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/30682 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/20558 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/30420 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/20558 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3288 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/33881 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/76698 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/32597 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/20558 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | |

### CODE REVIEW

**Comment by weinig** [comment]
Going to re-import css/css-color WPT first, as otherwise this is unreviewable. - https://bugs.webkit.org/show_bug.cgi?id=245348

File: Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp [hunk 1]
```diff
@@ -2184,6 +2184,123 @@ static Color parseHWBParameters(CSSParserTokenRange& range, const CSSParserConte
     return parseNonRelativeHWBParameters(args, context);
 }
 
+// The NormalizePercentage structs are specialized for the color types
+// that have specified rules for normalizing percentages by component.
+// The structs contains static members for the component that describe
+// the normalized value when the percent is 100%. As all treat 0% as
+// normalizing to 0, that is not encoded in the struct.
+template <typename ColorType> struct NormalizePercentage;
```
Comment (darinadler) [inline Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp]: Remove the space before "<"?

File: Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp [hunk 2]
```diff
@@ -2207,15 +2324,18 @@ static Color parseLabParameters(CSSParserTokenRange& args, ConsumerForLightness&
         return { };
 
     auto normalizedLightness = WTF::switchOn(*lightness,
-        [] (PercentRaw percent) { return std::max(0.0, percent.value); },
+        [] (NumberRaw number) { return  std::max(0.0, number.value); },
```
Comment (darinadler) [inline Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp]: Remove space after "return"?

**Review by darinadler — APPROVED** [review]
What's the story of the WPT tests? Frustrating to have this patch introduce so many new failures but I am sure you have thought about that.

File: Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp (hunk 2 above)
Comment (weinig) [reply Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp]: My absence has made me sloppy :(.

**Comment by weinig** [comment]
> What's the story of the WPT tests? Frustrating to have this patch introduce so many new failures but I am sure you have thought about that.

Going to have a followup change that updates the tests in parallel to working to get the tests updated in WPT.

**Pull request merged**